src/wallpaperController.ts

```diff
@@ -0,0 +1,646 @@
+            sourceResult.push(chosenSource);
+        }
+
+        return sourceResult;
+    }
+
+    /**
+     * Sets the wallpaper and the lock screen when enabled to the given path.
+     *
+     * Types:
+     * 0: Background
+     * 1: Lock screen
+     * 2: Background and lock screen
+     *
+     * @param {string[]} wallpaperPaths Array of paths to the image
+     * @param {number} type Types to change
+     */
+    private async _setBackground(wallpaperPaths: string[], type: number = 0) {
+        const backgroundSettings = new SettingsModule.Settings('org.gnome.desktop.background');
+        const screensaverSettings = new SettingsModule.Settings('org.gnome.desktop.screensaver');
+
+        if (wallpaperPaths.length < 1)
+            throw new Error('Empty wallpaper array');
+
+        const wallpaperUri = `file://${wallpaperPaths[0]}`;
+
+        if (wallpaperPaths.length > 1 && this._wallpaperManager) {
+            await this._wallpaperManager.setWallpaper(wallpaperPaths, type, backgroundSettings, screensaverSettings);
+            return;
+        }
+
+        if (type === 0 || type === 2) {
+            // FIXME: These are currently hardcoded for the few available wallpaperManager
+            if (wallpaperUri.includes('merged_wallpaper') || wallpaperUri.includes('cli-a') || wallpaperUri.includes('cli-b'))
+            // merged wallpapers need mode "spanned"
+                backgroundSettings.setString('picture-options', 'spanned');
+            else
+            // single wallpapers need mode "zoom"
+                backgroundSettings.setString('picture-options', 'zoom');
+
+            Utils.setPictureUriOfSettingsObject(backgroundSettings, wallpaperUri);
+        }
+
+        if (type === 1) {
+            // FIXME: These are currently hardcoded for the few available wallpaperManager
+            if (wallpaperUri.includes('merged_wallpaper') || wallpaperUri.includes('cli-a') || wallpaperUri.includes('cli-b'))
+            // merged wallpapers need mode "spanned"
+                screensaverSettings.setString('picture-options', 'spanned');
+            else
+            // single wallpapers need mode "zoom"
+                screensaverSettings.setString('picture-options', 'zoom');
+
+            Utils.setPictureUriOfSettingsObject(screensaverSettings, wallpaperUri);
+        }
+
+        if (type === 2) {
+            // FIXME: These are currently hardcoded for the few available wallpaperManager
+            if (wallpaperUri.includes('merged_wallpaper') || wallpaperUri.includes('cli-a') || wallpaperUri.includes('cli-b'))
+                // merged wallpapers need mode "spanned"
+                screensaverSettings.setString('picture-options', 'spanned');
+            else
+                // single wallpapers need mode "zoom"
+                screensaverSettings.setString('picture-options', 'zoom');
+
+            Utils.setPictureUriOfSettingsObject(screensaverSettings, backgroundSettings.getString('picture-uri'));
```

Comment:
Can you briefly explain why you have to set the wallpaper in the new wallpaper manager instance and here again? Maybe we can reduce the places where we actually set picture URIs?

Reply:
This is the original place from your old code. Initially I gave HydraPaper some special treatment inline but while adding SuperPaper I moved those to the according manager.
They behave slightly different in what settings they set on their own and how we have to handle these.

This code here has some special treatment as well for hover previews. When returning to the current wallpaper it could be an already merged wallpaper which then needs a different zoom option. That's the search in the name for 'merged_wallpaper' and 'cli-[a|b]'.

For reducing complexity I see the following options:
* We could introduce a new default wallpaperManager instance which would handle these default cases. Setting the URIs would be in similar places then.
* We could ditch the lockscreen setting. That would remove the need for special treatments. I honestly never saw that showing up anywhere. But the gnome settings app also sets the lockscreen value on wallpaper change so it should show up somewhere?

Reply:
I added the first option locally. Have to give it some tests before I'll push here tough.

Reply:
> I honestly never saw that showing up anywhere. But the gnome settings app also sets the lockscreen value on wallpaper change so it should show up somewhere?

I don't know if it is actually used anywhere in gnome anymore. It used to show up when locking/unlocking your screen iirc. Now there is always the blurred background I think... :shrug: